workers/unity/Packages/com.improbable.gdk.tools/DownloadCoreSdk.cs

```diff
@@ -1,53 +1,127 @@
+using System.Collections.Generic;
 using System.IO;
-using UnityEngine;
+using System.Linq;
 using UnityEditor;
+using UnityEngine;
 
 namespace Improbable.Gdk.Tools
 {
+    public enum DownloadResult
+    {
+        Success,
+        AlreadyInstalled,
+        Error
+    }
+
     public static class DownloadCoreSdk
     {
-        private const int DownloadPriority = 50;
-        private const int DownloadForcePriority = 51;
+        private const int DownloadForcePriority = 50;
 
-        [MenuItem("Improbable/Download CoreSdk", false, DownloadPriority)]
-        public static void DownloadMenu()
+        private static readonly string InstallMarkerFile =
+            Path.GetFullPath($"build/CoreSdk/{Common.CoreSdkVersion}.installed");
+
+        private static readonly string ProjectPath =
+            Path.GetFullPath(Path.Combine(Common.GetThisPackagePath(), ".DownloadCoreSdk/DownloadCoreSdk.csproj"));
+
+        private const string DownloadForceMenuItem = "Improbable/Download CoreSdk (force)";
+
+        [MenuItem(DownloadForceMenuItem, false, DownloadForcePriority)]
+        private static void DownloadForceMenu()
         {
             Download();
             AssetDatabase.Refresh();
         }
 
-        [MenuItem("Improbable/Download CoreSdk (force)", false, DownloadForcePriority)]
-        public static void DownloadForceMenu()
+        /// <summary>
+        ///     Windows only: Ensures that the user can't try to force a download if any of Improbable's native plugins are loaded
+        ///     by Unity.
+        /// </summary>
+        [MenuItem(DownloadForceMenuItem, true, DownloadForcePriority)]
+        private static bool DownloadForceValidate()
         {
-            var path = Path.Combine("build", "CoreSdk");
-            if (Directory.Exists(path))
+            var failedPlugins = new List<PluginImporter>();
+
+            // Unity never unloads native plugins. Detect them here and give a more useful error message.
```

Comment:
Afaict - this doesn't enable you to print an error message - just early out from the command executing. Are we able to have a custom message on validate failed?

Reply:
Ah yeah, this is a stale comment. I could also leave the command enabled and print an error message (what I originally did.)
I've gone back and forth between disabling and printing.